examples/npm-virtualenv-uwsgi-test-app/app.sh

```diff
@@ -28,7 +11,10 @@ fi
 echo "Testing UPGRADE_PIP_TO_LATEST=1 (set in .s2i/environment) ..."
 packages=("pip" "setuptools" "wheel")
 for pkg in ${packages[@]}; do
-  if [ $(get_latest_stable_version $pkg) != $(get_installed_version $pkg) ]; then
+  # grep returns exit code 1 if the output contains only one line starting
+  # with "Requirement already …" which means that the package is updated
+  python -m pip install -U --no-python-version-warning $pkg | grep -v "^Requirement already up-to-date: "
+  if [ $? -eq 0 ]; then
```

Comment:
Have you tested what happens if the update fails for any other reason?

Reply:
This step should not update anything. If the command will have anything else in the output and the grep won't fail, it means that:
1. there is a problem checking the versions from PyPI
2. the installed packages are not the latest versions

Reply:
I don't understand why it won't update anything?

Reply:
This command is here only to check that all packages in the virtual environment are up to date. The packages are actually updated in the `$UPGRADE_PIP_TO_LATEST` section in the assemble script. This is only a test. If all the packages are correctly updated by the pip command in the assemble script, this command does nothing and the test pass. Otherwise, this command does an update which is wrong and the test fails.

Reply:
Aha! So this test is run only after pip has already been upgraded to latest? Ok that makes sense.

So we have 3 cases for the lines 16 and 17:
1. command works and grep fails => the if doesn't succeed - GOOD
2. command works and grep succeeds => if does succeed and reports error - GOOD
3. command fails => if does not succeed and does not report error - BAD?

Or is there something I'm missing with case 3?

Reply:
> Aha! So this test is run only after pip has already been upgraded to latest? Ok that makes sense.

pip, setuptools, wheel, yes

> 
> So we have 3 cases for the lines 16 and 17:
> 
>     1. command works and grep fails => the if doesn't succeed - GOOD
>
>     2. command works and grep succeeds => if does succeed and reports error - GOOD
> 
>     3. command fails => if does not succeed and does not report error - BAD?
> 
> 
> Or is there something I'm missing with case 3?

If pip fails and produces anything unexpected on stdout, grep does not fail (because the output is not empty) and the failure is reported.

```
$ pip install six --index-url https://example.com  | grep -v "^Requirement already up-to-date: "
Looking in indexes: https://example.com
ERROR: Could not install packages due to an EnvironmentError: ('Received response with content-encoding: gzip, but failed to decode it.', error('Error -3 while decompressing data: incorrect header check',))
$ echo $?
0
```

We can make it better for cases when pip produces something only to stderr by redirecting stderr to stdout so the test will fail in more cases. Is this what you meant?

Reply:
Ah, I thought a failed command does not continue with a pipe. Maybe that's a setting in bash?

In any case, thanks for the explanation. In that case it looks good to me!

Reply:
I've added the mentioned redirection of stderr to stdout to make this more bulletproof.